Feat/start vm choose power state

benjamreis opened the pull request:
- Add `power_state`(default `Halted`) and `suspend_VDI`(default `Ref.null`) fields to create a VM in `Suspended` power state
   - Add fields `[last_booted_record (default:""), last_boot_CPU_flags (default:[])]` to make the VM created suspended be able to be restored.

benjamreis committed:
Enable to create a VM in `Suspended` state with a `suspend_VDI` set

* add to the `vm.create` method:
    * `power_state` with default value: `Halted`
    * `suspend_VDI` with default value: `Ref.null`
    * `last_booted_record` with default value: `""`
    * `last_boot_CPU_flags` with default value: `[]`
  The default behavior is not changed
* add `device` (default: '') and `currently_attached` (default: false) to the `VBD.create` method to allow the creation of an already plugged VBD. the new fields can only be used if the VM is suspended. Default behaviour is kept.
* add `currently_attached` (default: false) to the `VIF.create` method to allow the creation of an already plugged VIF. the new fields can only be used if the VM is suspended. Default behaviour is kept.

Signed-off-by: BenjiReis <benjamin.reis@vates.fr>

benjamreis commented on ocaml/idl/datamodel_vm.ml:
```diff
@@ -1088,8 +1088,11 @@ let power_behaviour =
+      ~versioned_params:
+          [{param_type=Ref _vm; param_name="self"; param_doc="The VM"; param_release=boston_release; param_default=None};
+           {param_type=Ref _vdi; param_name="value"; param_doc="The suspend VDI uuid"; param_release=boston_release; param_default=None};
+           {param_type=Bool; param_name="ignore_checksum"; param_doc="Ignore checksum or not."; param_release=tampa_release; param_default=Some (VBool false)}
```
Find correct param release.

julien-f commented on ocaml/idl/datamodel_vm.ml:
```diff
@@ -1442,6 +1445,7 @@ let set_NVRAM_EFI_variables = call ~flags:[`Session]
            field ~lifecycle:[Prototyped, rel_naples, ""] ~qualifier:StaticRO ~ty:(Map(String, String)) "NVRAM"
              ~default_value:(Some (VMap []))
              "initial value for guest NVRAM (containing UEFI variables, etc). Cannot be changed while the VM is running";
+           field ~qualifier:StaticRO ~ty:Bool ~default_value:(Some (VBool false)) "suspended" "creates the VM in Suspended state";
```
Wouldn't be better to add a `power_state` parameter instead and rejects if the value is not authorized (`Halted` or `Suspended`)?

johnelse replied:
Agreed. VM already has a `power_state` member. It would be better to make that `StaticRO` so that it becomes a constructor argument.

julien-f commented on ocaml/idl/datamodel_vm.ml:
```diff
@@ -1088,8 +1088,11 @@ let power_behaviour =
+      ~versioned_params:
+          [{param_type=Ref _vm; param_name="self"; param_doc="The VM"; param_release=boston_release; param_default=None};
+           {param_type=Ref _vdi; param_name="value"; param_doc="The suspend VDI uuid"; param_release=boston_release; param_default=None};
+           {param_type=Bool; param_name="ignore_checksum"; param_doc="Ignore checksum or not."; param_release=tampa_release; param_default=Some (VBool false)}
```
IMHO, it would be better to simply allow a `suspend_VDI` field in `VM.create` in case `power_state` is `Suspended`.

This would limit this special case at VM creation and would not require changing `set_suspend_VDI`.

johnelse replied:
Agreed. It should be an error to try and create a not-suspended VM with a non-null `suspend_VDI`, though.

johnelse commented on ocaml/idl/datamodel_vm.ml:
```diff
@@ -1379,16 +1379,16 @@ let set_NVRAM_EFI_variables = call ~flags:[`Session]
              "PCI_bus" "PCI bus path for pass-through devices";
            field  ~ty:(Map(String, String)) "other_config" "additional configuration" ~map_keys_roles:["pci", _R_POOL_ADMIN; ("folder",(_R_VM_OP));("XenCenter.CustomFields.*",(_R_VM_OP))];
-           field ~qualifier:DynamicRO ~ty:Int "domid" "domain ID (if available, -1 otherwise)";
+           field ~qualifier:StaticRO ~ty:Int ~default_value:(Some (VInt (-1L))) "domid" "domain ID (if available, -1 otherwise)";
```
Why does the domid need to become `StaticRO` ? Suspended VMs still have domid -1 IIRC. They definitely don't have an associated domain.

benjamreis replied:
Just checked and it seems you're right. I'll try without but I remembered differently. ^^'

benjamreis commented on ocaml/xapi/xapi_vm.ml:
```diff
@@ -505,6 +507,16 @@ let create ~__context ~name_label ~name_description
+  else begin
+    error "Bad power state for VM creation";
+  end;
+
```
I think in this error cases we should raise? or return a null ref?

edwintorok committed:
CA-337546: update to new Stunnel_cache API with Safe_resources support

This required moving the stunnel_processes cleanup function to where the
tunnel was opened (convert from [connect] to [with_connect] API).

Purging needs a slight change since we no longer raise Not_found,
but return None instead.

Retrying to open the connection over TCP also needs to be made only
if the connection itself fails, not if the function [f] fails,
hence the use of a fallback variable.

Signed-off-by: Edwin Török <edvin.torok@citrix.com>

lindig committed:
Revert "CA-337867: Expose 'scheduled_to_be_resident_on' to XAPI event"

This led to BVT failures because guest templates are affected.

This reverts commit d15082d2bde4f001f31cab244493e5631b4c25ac.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

robhoes committed:
Merge pull request #4115 from lindig/master

Revert "CA-337867: Expose 'scheduled_to_be_resident_on' to XAPI event"

robhoes committed:
CA-337867: Expose 'VM.scheduled_to_be_resident_on' field (take 2)

This reintroduces d15082d2, but now with a default value added, which is
necessary for imports of existing VM images.

Signed-off-by: Rob Hoes <rob.hoes@citrix.com>

edwintorok committed:
maintenance: fix build with workspaces

When using dune workspaces (e.g. having a `dune-workspace` file in a directory a level above)
the build fails:
```
Entering directory '/home/edwin-work/toolstack'
File "xapi/ocaml/util/dune", line 1, characters 0-163:
1 | (rule
2 |   (targets build_info.ml)
3 |   (deps %{workspace_root}/ocaml/util/gen_build_info_ml.sh)
4 |   (action (with-stdout-to %{targets} (bash "./gen_build_info_ml.sh")))
5 | )
Error: No rule found for ocaml/util/gen_build_info_ml.sh
```

Do not reference workspace_root, instead just reference the script
relative to the dune file.

Signed-off-by: Edwin Török <edvin.torok@citrix.com>

robhoes committed:
Merge pull request #4116 from robhoes/ca337867

CA-337867: Expose 'VM.scheduled_to_be_resident_on' field (take 2)

lindig committed:
Revert "CA-339526 make gc_compact call public"

This reverts commit d39fcf2994ef29df9e1545fe2a63d3ee15031473.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

edwintorok committed:
CA-337546: enable runtime warnings

Runtime warnings are low-level warnings about leaks detected when using
OCaml channels and Unix file descriptors.

These warnings are printed from within a finalizer, so printing it
anywhere else than stderr is complicated
(risks deadlocking because some mutexes might be held while the finalizer is running).

Thus don't throw away stderr output, log it instead.

Signed-off-by: Edwin Török <edvin.torok@citrix.com>

Commit:
maintenance: improve IMPORT_INCOMPATIBLE_VERSION error message

Signed-off-by: lippirk <ben.anson@citrix.com>

Commit:
Merge pull request #4118 from lippirk/improve-import-error-message

maintenance: improve IMPORT_INCOMPATIBLE_VERSION error message

lindig committed:
Revert "CA-334811 assign xapi version automatically"

This reverts commit b81af6e7321129af3bdb85a0f5ddb187b404b1bd.
This caused CA-340484 and we are reverting this for the moment.

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

edwintorok committed:
Revert "Revert "CA-334811 assign xapi version automatically""

This reverts commit 5c05f22a187908b6c832113803607b2758e95264.
We have found the bug in guest-templates-json: it was running on a
slave, when it shouldn't.

Try to restore the version number change and the stricter check

Signed-off-by: Edwin Török <edvin.torok@citrix.com>
